google/cloud/spanner/sql_statement.h

```diff
@@ -0,0 +1,72 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_SPANNER_SQL_STATEMENT_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_SPANNER_SQL_STATEMENT_H_
+
+#include "google/cloud/spanner/value.h"
+#include <string>
+#include <unordered_map>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+/**
+ * The SqlStatement class representis a SQL statement, that may or may not be
+ * parameterized.
+ *
+ * Parameter placeholders are specified by @<param name> in the SQL string.
+ * Values for parameters are a collection of std::pair<std::string,
+ * google::cloud:spanner::Value>.
+ * Example:
+ *   google::cloud::spanner::SqlStatement stmt("select * from scmods where
+ *   last_name = @param1", {{"param1", Value("Blues")}});
+ */
+class SqlStatement {
+ public:
+  /// Type alias for parameter collection.
+  using param_type = std::unordered_map<std::string, Value>;
+
+  /// Copy and move.
```

Comment:
Does this also support default construction? It seems like it could. It'd likely be equivalent to `SqlStatement("")` maybe?

Reply:
If we allowed an empty SqlStatement, would it also follow that we would want mutators for statement_ and params_?

Reply:
Not necessarily. `SqlStatement` already supports mutations via assignment. So a default constructor could be useful like:

```cc
SqlStatement sql;
...
sql = ComputeStatement(...);
```

Reply:
Not necessarily, you may still use the default constructor in something like this:

```C++
SqlStatement sql;
if (foo) { sql = ...; } else { sql = ...other-dots...; }
```

Reply:
Done